source/guides/maintenance/custom-app-container.rst

```diff
@@ -0,0 +1,402 @@
+published as `Docker images on Docker Hub`_. The source for these Docker images
+is available on `GitHub`_\*.
+
+In this guide, the phrase *Clear Linux\* OS container base images* is used to
+refer to the official `clearlinux image on Docker Hub`_. This guide describes how
+to edit |CL| container base images for the following situations:
+
+
+*	Customize an application image Dockerfile by adding a bundle
+*	Rebuild a published Clear Linux OS Dockerfile with desired version
+
+ 	*	Single-stage Clear Linux OS Dockerfiles
+	*	Multi-stage Clear Linux OS Dockerfiles
+
+*	Customize an application image at runtime
+
+
+Prerequisites
+*************
+
+*	Set up a functional Docker environment as described in :ref:`docker`.
+*	Download the |CL| microservice Dockerfile repo with the following command:
+
+ 	.. code-block:: bash
+
+ 		git clone https://github.com/clearlinux/dockerfiles.git
+
+Background
+**********
+
+Most of the images in the `clearlinux microservice dockerfile repo`_ are built
+utilizing a Docker feature called a `multi-stage build`_:
+
+#.	clearlinux/os-core is built once a day.
+
+#.	The target takes clearlinux/os-core as base layer, via
+	:command:`FROM clearlinux/os-core:latest` or via
+	:command:`FROM clearlinux/<Clear Linux Service>:latest`, in which
+	:command:`clearlinux/<Clear Linux Service>` takes clearlinux/os-core as the base layer.
+
+#.	The specific features are built in the first stage via
+	:command:`swupd os-install … bundles=httpd…` with all files from its base layer removed.
+
+#.	The target is a composition of its base layer and the specific feature layer,
+	via :command:`FROM clearlinux/<base layer>:latest`, such as: os-core, httpd,
+	and via :command:`COPY --from=builder /install_root /`.
+
+Using this method, the target container images are kept up to date without
+file duplication. For application-centric containers, `os-core-update` is
+excluded to improve size optimization.
+
+Multi-stage Dockerfiles contain more than one :command:`FROM` directive. All
+of the multi-stage Clear Linux OS Dockerfiles share a common base layer called
+:command:`clearlinux/os-core:latest`. All of the higher level layers inherit
+the Clear Linux OS version from this base layer.
+
+Examples
+********
+
+This section describes the following examples for customizing an application image
+Dockerfile:
+
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Add a bundle
+============
+
+To add new software to a container image, you need to find out which |CL|
+bundle includes it, using the :command:`swupd search <feature>` command.
+(Refer to `swupd source documentation`_ on GitHub for command details.)
+
+After finding out the right bundle(s), update the Dockerfile to include the
+bundle into the published |CL| container. In this example, we chose
+to add :command:`wget` to the clearlinux/redis Dockerfile.
+
+To demonstrate this, a diff of a modified clearlinux/redis Dockerfile is

```

Comment:
The order of ops is not clear.  Users need a contextual clue: Where does this example start? This just needs a transition statement. There's a gap between Prerequisites (e.g., cloning the microservice Dockerfile repo) and where this example begins. It would be useful to split up the long code-block into at least a couple code-blocks. 
Please transform and simplify these paras into steps `#.` in this section.